src/nvim/plines.c

```diff
@@ -991,25 +994,43 @@ int64_t win_text_height(win_T *const wp, const linenr_T start_lnum, const int64_
     lnum = lnum_next + 1;
   }
 
-  while (lnum <= end_lnum) {
+  while (lnum <= *end_lnum && height_sum_nofill + height_sum_fill < max) {
     linenr_T lnum_next = lnum;
-    hasFolding(wp, lnum, &lnum, &lnum_next);
-    height_sum_fill += win_get_fill(wp, lnum);
+    cur_folded = hasFolding(wp, lnum, &lnum, &lnum_next);
+    fill_end = win_get_fill(wp, lnum);
+    height_sum_fill += fill_end;
     height_cur_nofill = plines_win_nofill(wp, lnum, false);
     height_sum_nofill += height_cur_nofill;
+    cur_lnum = lnum;
     lnum = lnum_next + 1;
   }
 
-  if (end_vcol >= 0) {
+  int64_t vcol_end = *end_vcol;
+  bool use_vcol = vcol_end >= 0 && lnum > *end_lnum;
+  if (use_vcol) {
     height_sum_nofill -= height_cur_nofill;
-    const int64_t row_off = end_vcol == 0
+    const int64_t row_off = vcol_end == 0
                             ? 0
-                            : (end_vcol <= width1 || width2 <= 0)
+                            : (vcol_end <= width1 || width2 <= 0)
                             ? 1
-                            : 1 + (end_vcol - width1 + width2 - 1) / width2;
+                            : 1 + (vcol_end - width1 + width2 - 1) / width2;
     height_sum_nofill += MIN(row_off, height_cur_nofill);
   }
 
+  if (cur_folded) {
+    vcol_end = 0;
+  } else {
+    int linesize = linetabsize_eol(wp, cur_lnum);
+    vcol_end = MIN(use_vcol ? vcol_end : INT64_MAX, linesize);
+  }
+
+  int64_t overflow = height_sum_nofill + height_sum_fill - fill_end - max;
```

Comment:
Don't remember why I subtracted `fill_end` here. Pushing an update that removes it with a test result that seems to make sense.